### PR TITLE
More Linux-like fs ls on non-windows

### PR DESCRIPTION
**Pull request by zanderz**

### CODE REVIEW

**Comment by maxtaco** [comment]
can you rebase before I review? (seems like there are merge conflicts)

Also, took a quick look at the ls library that's been excerpted. The style is going to get wrecked by our linter.  Was that library a quick port from C?

**Comment by maxtaco** [comment]
Just checked it out and ran it, I didn't get any output as far as I could tell:

```
[max@piave ~/src/go/src/github.com/keybase/client/go/keybase ] ~/src/go/bin/keybase --run-mode=prod fs ls /keybase/private/max
```

For Chris's public folder, I got something weird:

```
[max@piave ~/src/go/src/github.com/keybase/client/go/keybase ] ~/src/go/bin/keybase --run-mode=prod fs ls /keybase/public/chris

._.DS_Store  ._keybase.txt  .DS_Store  index.md  keybase.txt  keys  photos  plan.txt  verified_software

```

Note the two blank lines...

**Comment by zanderz** [comment]
Sorry, jumped the gun on that, please retry.

**Comment by maxtaco** [comment]
I ran `golint` on that new file, and it's not pretty

```
golint cmd_simplefs_list_nix.go
cmd_simplefs_list_nix.go:27:2: don't use underscores in Go names; const color_fg_black should be colorFgBlack
cmd_simplefs_list_nix.go:28:2: don't use underscores in Go names; const color_fg_red should be colorFgRed
cmd_simplefs_list_nix.go:29:2: don't use underscores in Go names; const color_fg_green should be colorFgGreen
cmd_simplefs_list_nix.go:30:2: don't use underscores in Go names; const color_fg_brown should be colorFgBrown
cmd_simplefs_list_nix.go:31:2: don't use underscores in Go names; const color_fg_blue should be colorFgBlue
cmd_simplefs_list_nix.go:32:2: don't use underscores in Go names; const color_fg_magenta should be colorFgMagenta
cmd_simplefs_list_nix.go:33:2: don't use underscores in Go names; const color_fg_cyan should be colorFgCyan
cmd_simplefs_list_nix.go:34:2: don't use underscores in Go names; const color_fg_white should be colorFgWhite
cmd_simplefs_list_nix.go:35:2: don't use underscores in Go names; const color_bg_black should be colorBgBlack
cmd_simplefs_list_nix.go:36:2: don't use underscores in Go names; const color_bg_red should be colorBgRed
cmd_simplefs_list_nix.go:37:2: don't use underscores in Go names; const color_bg_green should be colorBgGreen
cmd_simplefs_list_nix.go:38:2: don't use underscores in Go names; const color_bg_brown should be colorBgBrown
cmd_simplefs_list_nix.go:39:2: don't use underscores in Go names; const color_bg_blue should be colorBgBlue
cmd_simplefs_list_nix.go:40:2: don't use underscores in Go names; const color_bg_magenta should be colorBgMagenta
cmd_simplefs_list_nix.go:41:2: don't use underscores in Go names; const color_bg_cyan should be colorBgCyan
cmd_simplefs_list_nix.go:42:2: don't use underscores in Go names; const color_bg_white should be colorBgWhite
cmd_simplefs_list_nix.go:45:1: comment on exported type FileInfoPath should be of the form "FileInfoPath ..." (with optional leading article)
cmd_simplefs_list_nix.go:53:1: comment on exported type DirentFileInfo should be of the form "DirentFileInfo ..." (with optional leading article)
cmd_simplefs_list_nix.go:58:1: exported method DirentFileInfo.Name should have comment or be unexported
cmd_simplefs_list_nix.go:63:1: exported method DirentFileInfo.Size should have comment or be unexported
cmd_simplefs_list_nix.go:68:1: exported method DirentFileInfo.Mode should have comment or be unexported
cmd_simplefs_list_nix.go:83:1: exported method DirentFileInfo.ModTime should have comment or be unexported
cmd_simplefs_list_nix.go:88:1: exported method DirentFileInfo.IsDir should have comment or be unexported
cmd_simplefs_list_nix.go:96:1: exported method DirentFileInfo.Sys should have comment or be unexported
cmd_simplefs_list_nix.go:101:1: comment on exported type Listing should be of the form "Listing ..." (with optional leading article)
cmd_simplefs_list_nix.go:105:2: don't use underscores in Go names; struct field num_hard_links should be numHardLinks
cmd_simplefs_list_nix.go:109:2: don't use underscores in Go names; struct field epoch_nano should be epochNano
cmd_simplefs_list_nix.go:114:2: don't use underscores in Go names; struct field link_name should be linkName
cmd_simplefs_list_nix.go:115:2: don't use underscores in Go names; struct field link_orphan should be linkOrphan
cmd_simplefs_list_nix.go:116:2: don't use underscores in Go names; struct field is_socket should be isSocket
cmd_simplefs_list_nix.go:117:2: don't use underscores in Go names; struct field is_pipe should be isPipe
cmd_simplefs_list_nix.go:118:2: don't use underscores in Go names; struct field is_block should be isBlock
cmd_simplefs_list_nix.go:119:2: don't use underscores in Go names; struct field is_character should be isCharacter
cmd_simplefs_list_nix.go:124:2: don't use underscores in Go names; var color_map should be colorMap
cmd_simplefs_list_nix.go:130:6: don't use underscores in Go names; func get_partial_color should be getPartialColor
cmd_simplefs_list_nix.go:131:6: don't use underscores in Go names; var partial_bytes should be partialBytes
cmd_simplefs_list_nix.go:216:6: don't use underscores in Go names; func get_color_from_bsd_code should be getColorFromBsdCode
cmd_simplefs_list_nix.go:217:2: don't use underscores in Go names; var color_foreground should be colorForeground
cmd_simplefs_list_nix.go:218:2: don't use underscores in Go names; var color_background should be colorBackground
cmd_simplefs_list_nix.go:220:6: don't use underscores in Go names; var color_bytes should be colorBytes
cmd_simplefs_list_nix.go:231:6: don't use underscores in Go names; func parse_LSCOLORS should be parseLSCOLORS
cmd_simplefs_list_nix.go:272:27: don't use underscores in Go names; method write_listing_name should be writeListingName
cmd_simplefs_list_nix.go:272:46: don't use underscores in Go names; method parameter output_buffer should be outputBuffer
cmd_simplefs_list_nix.go:275:3: don't use underscores in Go names; var applied_color should be appliedColor
cmd_simplefs_list_nix.go:277:3: don't use underscores in Go names; var num_hardlinks should be numHardlinks
cmd_simplefs_list_nix.go:280:3: don't use underscores in Go names; var name_split should be nameSplit
cmd_simplefs_list_nix.go:281:3: don't use underscores in Go names; var extension_str should be extensionStr
cmd_simplefs_list_nix.go:356:27: don't use underscores in Go names; method create_listing should be createListing
cmd_simplefs_list_nix.go:357:6: don't use underscores in Go names; var current_listing should be currentListing
cmd_simplefs_list_nix.go:378:7: don't use underscores in Go names; var _link_pathstr should be _linkPathstr
cmd_simplefs_list_nix.go:456:3: don't use underscores in Go names; var size_str should be sizeStr
cmd_simplefs_list_nix.go:458:4: don't use underscores in Go names; var size_b should be sizeB
cmd_simplefs_list_nix.go:490:2: don't use underscores in Go names; var epoch_now should be epochNow
cmd_simplefs_list_nix.go:491:6: don't use underscores in Go names; var seconds_in_six_months should be secondsInSixMonths
cmd_simplefs_list_nix.go:492:2: don't use underscores in Go names; var epoch_six_months_ago should be epochSixMonthsAgo
cmd_simplefs_list_nix.go:493:2: don't use underscores in Go names; var epoch_modified should be epochModified
cmd_simplefs_list_nix.go:495:6: don't use underscores in Go names; var time_str should be timeStr
cmd_simplefs_list_nix.go:524:6: don't use underscores in Go names; func compare_name should be compareName
cmd_simplefs_list_nix.go:525:2: don't use underscores in Go names; var a_name_lower should be aNameLower
cmd_simplefs_list_nix.go:526:2: don't use underscores in Go names; var b_name_lower should be bNameLower
cmd_simplefs_list_nix.go:528:6: don't use underscores in Go names; var smaller_len should be smallerLen
cmd_simplefs_list_nix.go:554:6: don't use underscores in Go names; func compare_time should be compareTime
cmd_simplefs_list_nix.go:564:6: don't use underscores in Go names; func compare_size should be compareSize
cmd_simplefs_list_nix.go:565:2: don't use underscores in Go names; var a_size should be aSize
cmd_simplefs_list_nix.go:566:2: don't use underscores in Go names; var b_size should be bSize
cmd_simplefs_list_nix.go:576:27: don't use underscores in Go names; method sort_listings should be sortListings
cmd_simplefs_list_nix.go:577:2: don't use underscores in Go names; var comparison_function should be comparisonFunction
cmd_simplefs_list_nix.go:603:3: don't use underscores in Go names; var middle_index should be middleIndex
cmd_simplefs_list_nix.go:609:4: don't use underscores in Go names; var front_index should be frontIndex
cmd_simplefs_list_nix.go:610:4: don't use underscores in Go names; var rear_index should be rearIndex
cmd_simplefs_list_nix.go:625:27: don't use underscores in Go names; method write_listings_to_buffer should be writeListingsToBuffer
cmd_simplefs_list_nix.go:625:52: don't use underscores in Go names; method parameter output_buffer should be outputBuffer
cmd_simplefs_list_nix.go:627:2: don't use underscores in Go names; method parameter terminal_width should be terminalWidth
cmd_simplefs_list_nix.go:635:4: don't use underscores in Go names; var width_permissions should be widthPermissions
cmd_simplefs_list_nix.go:635:31: should drop = 0 from declaration of var width_permissions; it is the zero value
cmd_simplefs_list_nix.go:636:4: don't use underscores in Go names; var width_num_hard_links should be widthNumHardLinks
cmd_simplefs_list_nix.go:636:31: should drop = 0 from declaration of var width_num_hard_links; it is the zero value
cmd_simplefs_list_nix.go:637:4: don't use underscores in Go names; var width_owner should be widthOwner
cmd_simplefs_list_nix.go:637:31: should drop = 0 from declaration of var width_owner; it is the zero value
cmd_simplefs_list_nix.go:638:4: don't use underscores in Go names; var width_group should be widthGroup
cmd_simplefs_list_nix.go:638:31: should drop = 0 from declaration of var width_group; it is the zero value
cmd_simplefs_list_nix.go:639:4: don't use underscores in Go names; var width_size should be widthSize
cmd_simplefs_list_nix.go:639:31: should drop = 0 from declaration of var width_size; it is the zero value
cmd_simplefs_list_nix.go:640:4: don't use underscores in Go names; var width_time should be widthTime
cmd_simplefs_list_nix.go:640:31: should drop = 0 from declaration of var width_time; it is the zero value
cmd_simplefs_list_nix.go:734:3: don't use underscores in Go names; var num_rows should be numRows
cmd_simplefs_list_nix.go:735:7: don't use underscores in Go names; var col_widths should be colWidths
cmd_simplefs_list_nix.go:737:4: don't use underscores in Go names; var num_cols_float should be numColsFloat
cmd_simplefs_list_nix.go:739:4: don't use underscores in Go names; var num_cols should be numCols
cmd_simplefs_list_nix.go:742:11: should omit 2nd value from range; this loop is equivalent to `for i := range ...`
cmd_simplefs_list_nix.go:746:4: don't use underscores in Go names; var col_listings should be colListings
cmd_simplefs_list_nix.go:762:4: don't use underscores in Go names; var max_row_length should be maxRowLength
cmd_simplefs_list_nix.go:764:5: don't use underscores in Go names; var max_row_length should be maxRowLength
cmd_simplefs_list_nix.go:766:4: don't use underscores in Go names; var max_row_length should be maxRowLength
cmd_simplefs_list_nix.go:773:5: don't use underscores in Go names; var listings_in_first_col should be listingsInFirstCol
cmd_simplefs_list_nix.go:774:5: don't use underscores in Go names; var listings_in_last_col should be listingsInLastCol
cmd_simplefs_list_nix.go:806:30: don't use underscores in Go names; method parameter output_buffer should be outputBuffer
cmd_simplefs_list_nix.go:807:2: don't use underscores in Go names; var list_dirs should be listDirs
cmd_simplefs_list_nix.go:808:2: don't use underscores in Go names; var list_files should be listFiles
cmd_simplefs_list_nix.go:818:3: don't use ALL_CAPS in Go names; use CamelCase
cmd_simplefs_list_nix.go:825:4: don't use underscores in Go names; var LS_COLORS_split should be LSCOLORSSplit
cmd_simplefs_list_nix.go:831:5: don't use underscores in Go names; var i_split should be iSplit
cmd_simplefs_list_nix.go:832:5: don't use underscores in Go names; var color_code should be colorCode
cmd_simplefs_list_nix.go:887:3: don't use underscores in Go names; var f_listing should be fListing
cmd_simplefs_list_nix.go:906:2: don't use underscores in Go names; var num_files should be numFiles
cmd_simplefs_list_nix.go:907:2: don't use underscores in Go names; var num_dirs should be numDirs
cmd_simplefs_list_nix.go:947:2: don't use underscores in Go names; var terminal_width should be terminalWidth
cmd_simplefs_list_nix.go:952:6: don't use underscores in Go names; var output_buffer should be outputBuffer
```

**Comment by maxtaco** [comment]
Output looks great! One bug I noted: I can't ls my home dir `/keybase/private/max/`. Everything else works as expected

**Comment by maxtaco** [comment]
BTW, `fs ls` runs fine on my home dir with what's in master, so there's some failure as a result of this PR.

Sure, the `-w` flag is fine with me.

Here's our current rule on the linting:

```  output="$(golint "$file" 2>&1 | grep -v ^go/vendor | grep -v ^go/protocol | grep -v comment | grep -v KeysById | grep -v "error should be the last type" | grep -v stutters | grep -v "error strings should not be capitalized" | tee /dev/fd/5)"```

It would be a shame to lose the "no snake case" rule because that's actually something we enforce elsewhere.

**Comment by zanderz** [comment]
I guess the new commit hook has diverged. Maybe I'll spend some more time
trying to make it work on Windows again.

**Comment by maxtaco** [comment]
No debug or error output. Seems like something somewhere is silently dying on an error.

**Comment by zanderz** [comment]
can you try it again with these changes? It wasn't returning the error from output(). Thanks.

**Comment by maxtaco** [comment]
now I see: no output, no error, but the extra newline after the display has come back

**Comment by maxtaco** [comment]
(when I run ls on a directory that previously did work)

**Comment by zanderz** [comment]
I'd like to try again - should I come and do it?

**Comment by zanderz** [comment]
There is still an extra newline added, I think, by Keybase after every command. If this is too many I can change it to try to eat the last newline.

**Comment by maxtaco** [comment]
I don't think keybase puts in a newline after every command.

File: go/client/cmd_simplefs_list.go [hunk 1]
```diff
@@ -95,28 +198,58 @@ func (c *CmdSimpleFSList) Run() error {
...
+		for _, e := range listResult.Entries {
+			if e.DirentType == keybase1.DirentType_DIR {
+				ui.Printf("%s\t<%s>\t\t%s\n", keybase1.FormatTime(e.Time), keybase1.DirentTypeRevMap[e.DirentType], e.Name)
+			} else {
+				ui.Printf("%s\t%s\t%d\t%s\n", keybase1.FormatTime(e.Time), keybase1.DirentTypeRevMap[e.DirentType], e.Size, e.Name)
+			}
+		}
+	} else {
+		// capture the current terminal dimensions
+		terminalWidth, _, err := terminal.GetSize(int(os.Stdout.Fd()))
```
Comment (maxtaco) [inline go/client/cmd_simplefs_list.go]: Don't use this, we already have support in `UI` for getting the terminal size.

File: go/client/cmd_simplefs_list.go [hunk 2]
```diff
@@ -95,28 +198,58 @@ func (c *CmdSimpleFSList) Run() error {
...
+
+		var outputBuffer bytes.Buffer
+
+		err = c.ls(&outputBuffer, listResult, terminalWidth)
+		if err != nil {
+			return err
+		}
+
+		if outputBuffer.String() != "" {
+			ui.Printf("%s", outputBuffer.String())
```
Comment (maxtaco) [inline go/client/cmd_simplefs_list.go]: my guess is this `outputBuffer` has two trailing newlines.  can you check?

Comment (zanderz) [inline go/client/cmd_simplefs_list.go]: I did check, it does not. I made this `ui.Printf("%s -- end", outputBuffer.String())` and got:
```
c:\work\src\github.com\keybase\client\go\keybase>keybase fs ls -l /keybase/private/zanderz
-rw-rw-r--      116460896 Mar 03 10:47 Keybase_1.0.18-20170302120457+22e9a36.386.exe
-rw-rw-r--        4097907 Feb 27 10:35 Peter Berkow - Practice 2 - Purple Hair.mp3
-rw-rw-r--            736 Feb 08 18:16 setup.bat
-rw-rw-r--            436 Feb 08 18:16 sshnotes.txt
drw-rw-r--            920 Mar 08 10:24 test
-rw-rw-r--             27 Feb 28 13:10 test4.txt
-rw-rw-r--            618 Feb 08 18:16 testkeys.txt
 -- end
c:\work\src\github.com\keybase\client\go\keybase>
```

so it really seems like Keybase itself adds a newline. But we could shave off the last newline if you want.

Comment (maxtaco) [inline go/client/cmd_simplefs_list.go]: but i don't know what would be adding the extra newline and none of the other commands have this!

File: go/client/cmd_simplefs_list.go [hunk 3]
```diff
@@ -95,28 +198,58 @@ func (c *CmdSimpleFSList) Run() error {
 		if err != nil {
 			return err
 		}
+		gotList := false
 		for {
 			listResult, err := cli.SimpleFSReadList(ctx, opid)
+			// Eat the error here because it may just mean the results
```
Comment (maxtaco) [inline go/client/cmd_simplefs_list.go]: my guess is that `listResult` is empty the last time through this loop. And you should exit there. My guess is that solves two problems: (1) it simplifies the exit code below (an error is always an error and it allows you to get rid of the TODO); and (2) it explains where the empty line is coming from, which is running `output` on an empty list.

**Comment by zanderz** [comment]
I think you built off an older commit possibly? I see no extra newline in linux, even without checking the number of results.

**Comment by maxtaco** [comment]
try a directory with >50 entries or so, like my home dir

**Comment by maxtaco** [comment]
ok, tested, it works....

btw, i think the right interface here (as the todo hints) is to give an empty result at the end of the list, and not an error. we'd have to change that with @taruti

**Pull request merged**